_posts/2019-11-07-tfp-cran/tfp-cran.Rmd

```diff
@@ -0,0 +1,252 @@
+
+```{r}
+pred_dists <- model(x_test)
+pred_means <- pred_dists %>% tfd_mean()
+```
+
+You may be wondering, what good is this? One way or the other, we'll decide on picking the class with the highest probability, right?
+
+Right, but there are a number of interesting things you can do with these layers.
+We'll quickly introduce three well-known ones here, but wouldn't be surprised if we saw a lot more emerging in the near future.
+
+### Variational autoencoders, the elegant way
+
+Variational autoencoders are a prime example of something that got way easier to code when TF-2 style custom models and custom training appeared (TF-2 _style_, not TF-2, as these techniques actually became available more than a year before TF 2 was finally released). ^[See [Representation learning with MMD-VAE](https://blogs.rstudio.com/tensorflow/posts/2018-10-22-mmd-vae/) for an example.]
+
+Evolutionarily, the next step was to use TFP distributions ^[Done so, for example, in [Getting started with TensorFlow Probability from R](https://blogs.rstudio.com/tensorflow/posts/2019-01-08-getting-started-with-tf-probability/).], but back in the time some fiddling was required, as distributions could not yet alias as layers.
+
+Due to those hybrids though, we now can do something like this ^[For a complete running example, see the [tfprobability README](https://rstudio.github.io/tfprobability/).]:
+
+```{r}
+# encoder
+encoder_model <- keras_model_sequential() %>%
+  [...] %>%
+  layer_multivariate_normal_tri_l(event_size = encoded_size) %>%
+  layer_kl_divergence_add_loss([...])
+
+# decoder
+decoder_model <- keras_model_sequential() %>%
+  [...] %>%
+ layer_independent_bernoulli([...])
+
+# complete VAE
+vae_model <- keras_model(inputs = encoder_model$inputs,
+                         outputs = decoder_model(encoder_model$outputs[1]))
+
+# loss function
+vae_loss <- function (x, rv_x) - (rv_x %>% tfd_log_prob(x))
+```
+
+Both the encoder and the decoder are "just" sequential models, joined through the functional API(`keras_model`). The loss is just the negative log-probability of the data given the model. So where is the other part of the (negative) ELBO, the KL divergence? It is implicit in the encoder's output layer, [layer_kl_divergence_add_loss](https://rstudio.github.io/tfprobability/reference/layer_kl_divergence_add_loss.html). 
+
+Our other two examples involve quantifying uncertainty.
```

Comment:
```suggestion
Our two other examples involve quantifying uncertainty.
```

Reply:
There's no _true_ rule on the order of adjectives in English (some [blogs say otherwise](https://www.gingersoftware.com/content/grammar-rules/adjectives/order-of-adjectives/) -- yay internet), but having number/quantity first just sounds better to me. _Sprachgefühl!_

Reply:
Bonus nerd finding: Mark Forsyth (whose books of words I love) claims it's _opinion-size-age-shape-colour-origin-material-purpose Noun_.

Reply:
1) d'accord :-)
2) yeah, i knew that one... should check out the books!